iothub_client/samples/pnp/common/pnp_protocol_helpers.c

```diff
@@ -0,0 +1,327 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// Header associated with this .c file
+#include "pnp_protocol_helpers.h"
+
+// JSON parsing library
+#include "parson.h"
+
+// IoT core utility related header files
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/strings.h"
+
+// Format used when building a response for a root property that does not contain metadata
+static const char g_propertyWithoutResponseSchemaWithoutComponent[] = "{\"%s\":%s}";
+// Format used when building a response for a component's property that does not contain metadata
+static const char g_propertyWithoutResponseSchemaWithComponent[] = "{\"""%s\":{\"__t\":\"c\",\"%s\":%s}}";
+// Format used when building a response for a root property that does contain metadata
+static const char g_propertyWithResponseSchemaWithoutComponent[] =  "{\"%s\":{\"value\":%s,\"ac\":%d,\"ad\":\"%s\",\"av\":%d}}";
+// Format used when building a response for a component's property that does contain metadata
+static const char g_propertyWithResponseSchemaWithComponent[] =  "{\"""%s\":{\"__t\":\"c\",\"%s\":{\"value\":%s,\"ac\":%d,\"ad\":\"%s\",\"av\":%d}}}";
+
+// Character that separates a PnP component from the specific command on the component.
+static const char g_commandSeparator = '*';
+
+// The version of the desired twin is represented by the $version metadata.
+static const char g_JSONPropertyVersion[] = "$version";
+
+// IoTHub adds a JSON field "__t":"c" into desired top-level JSON objects that represent components.  Without this marking, the object 
+// is treated as a property off the root component.
+static const char g_JSONComponentMetadata[] = "__t";
+
+// Name of desired JSON field when retrieving a full twin.
+static const char g_JSONDesiredName[] = "desired";
+
+// Telemetry message property used to indicate the message's component.
+static const char PnP_TelemetryComponentProperty[] = "$.sub";
+
+STRING_HANDLE PnPHelper_CreateReportedProperty(const char* componentName, const char* propertyName, const char* propertyValue)
+{
+    STRING_HANDLE jsonToSend;
+
+    if (componentName == NULL) 
+    {
+        jsonToSend = STRING_construct_sprintf(g_propertyWithoutResponseSchemaWithoutComponent, propertyName, propertyValue);
```

Comment:
```suggestion
        jsonToSend = STRING_construct_sprintf(componentName  ? g_propertyWithoutResponseSchemaWithComponent : g_propertyWithoutResponseSchemaWithoutComponent, propertyName, propertyValue);
```

Reply:
The issue with this is that `g_propertyWithoutResponseSchemaWithComponent` takes 3 variadic whereas `g_propertyWithoutResponseSchemaWithoutComponent` only takes 2.  I guess I could put some sort of dummy %s with a "" in if componentName==NULL, but I don't think that's as readable as having separate statements.  So I plan to leave as-is.